File: sampler/dDNNF.py
import copy

def grow(l, n):
    while len(l) < n:
        l.append(0)

def deep_cp(l):
    return copy.deepcopy(l)

def right_shift(l, n):
    for i in range(0, n):
        l.insert(0, 0)
        if l[len(l) - 1] == 0:
            l.pop()

def add(l1, l2):
    grow(l1, len(l2))

    for i in range(0, len(l2)):
        l1[i] += l2[i]

def mult(l, scal):
    for i in range(0, len(l)):
        l[i] *= scal

def prod(l1, l2):
    res = []

    for i in range(0, len(l2)):
        if l2[i] != 0:
            tmp = deep_cp(l1)
            right_shift(tmp, i)
            mult(tmp, l2[i])
            add(res, tmp)
    return res



class Edge:
    def __init__(self, target, consts, free):
        self.target = target
        self.consts = consts
        self.free = free
        self.mc = 0

class Node:
    def __init__(self, num):
        self.num = num
        self.mc = -1
        self.mc_by_var = {}
        self.mc_by_nb_vars = []

    def add_child(self, target, consts, free):
        pass

    def annotate_mc(self, assumps):
        pass

    def get_childrend_ids(self):
        pass

    def get_vars(self):
        pass

    def print_dot(self):
        pass

class AndNode(Node):
    def __init__(self, num):
        Node.__init__(self, num)
        self.children = []

    def add_child(self, target, consts, free):
        if len(consts) != 0 or len(free) != 0:
            raise ValueError('AndNode children do not have consts or free variables')

        self.children.append(Edge(target, consts, free))

    def annotate_mc(self, assumps):
        self.mc = 1

        if len(self.children) == 0:
            self.mc = 0

        for i in self.children:
            self.mc *= i.target.mc
            i.mc = i.target.mc

        if len(self.children) == 0:
            self.mc = 0
            # self.mc_by_var = {0: 0}
            # self.mc_by_nb_vars = [0]
        else:
            # self.mc_by_var[0] = self.mc
            pass



    def get_childrend_ids(self):
        return [c.target.num for c in self.children]

    def get_vars(self):
        res = set()

        for c in self.children:
            res.update(c.free)
            res.update([abs(x) for x in c.consts])
        return res

    def print_dot(self):
        print(f"\"{self.num}\" [label=\"A {self.num}: {self.mc}\"];")


class OrNode(Node):
    def __init__(self, num):
        Node.__init__(self, num)
        self.children = []

    def add_child(self, target, consts, free):
        self.children.append(Edge(target, consts, free))

    def annotate_mc(self, assumps):
        self.mc = 0
        # self.mc_by_nb_vars = [0]

        for i in self.children:
            lmc = i.target.mc

            for l in i.consts:
                if -1 * l in assumps:
                    lmc = 0

            for v in i.free:
                if v not in assumps and -1 * v not in assumps:
                    lmc *= 2

            self.mc += lmc
            i.mc = lmc
            # self.mc += i.target.mc * (2**len(i.free))

        if len(self.children) == 0:
            self.mc = 1
            # self.mc_by_var = {0: 1}
        else:
            # self.mc_by_var = buf[0]
            pass


    def get_childrend_ids(self):
        return [c.target.num for c in self.children]

    def get_vars(self):
        res = set()

        for c in self.children:
            res.update(c.free)
            res.update([abs(x) for x in c.consts])
        return res

    def print_dot(self):
        print(f"\"{self.num}\" [label=\"O {self.num}: {self.mc}\"];")

class UnaryNode(Node):
    def __init__(self, num):
        Node.__init__(self, num)
        self.child = None

    def add_child(self, target, consts, free):
        self.child = Edge(target, consts, free)

    def annotate_mc(self, assumps):
        # self.mc = self.child.target.mc * (2**len(self.child.free))
        lmc = self.child.target.mc

        for l in self.child.consts:
            if -1 * l in assumps:
                lmc = 0

        for v in self.child.free:
            if v not in assumps and -1 * v not in assumps:
                lmc *= 2

        self.mc = lmc
        self.child.mc = lmc

    def get_childrend_ids(self):
        return [self.child.target.num]

    def get_vars(self):
        tmp = set([abs(x) for x in self.child.consts])
        tmp.update(self.child.free)
        return tmp

    def print_dot(self):
        print(f"\"{self.num}\" [label=\"U {self.num}: {self.mc}\"];")

class TrueNode(Node):
    def __init__(self, num):
        Node.__init__(self, num)
        self.mc = 1

    def add_child(self, target, consts, free):
        raise ValueError('Cannot add child to TrueNode')

    def annotate_mc(self, assumps):
        self.mc = 1
        # self.mc_by_var = {0: 1}
        # self.mc_by_nb_vars = [1]

    def get_childrend_ids(self):
        return []

    def get_vars(self):
        return set()

    def print_dot(self):
        print(f"\"{self.num}\" [label=\"T {self.num}: {self.mc}\"];")


class FalseNode(Node):
    def __init__(self, num):
        Node.__init__(self, num)
        self.mc = 0

    def add_child(self, target, consts, free):
        raise ValueError('Cannot add child to FalseNode')

    def annotate_mc(self, assumps):
        self.mc = 0
        # self.mc_by_var = {0: 0}
        # self.mc_by_nb_vars = [0]

    def get_childrend_ids(self):
        return []

    def get_vars(self):
        return set()

    def print_dot(self):
        print(f"\"{self.num}\" [label=\"F {self.num}: {self.mc}\"];")

class dDNNF:
    def __init__(self):
        self.nodes = []
        self.ordering = []

    def add_node(self, num, c):
        numt = len(self.nodes) + 1
        if num != numt:
            raise ValueError(f'node id {num} and calculated id {numt} do not match')

        self.nodes.append(c(num))
        return num

    def get_node(self, num):
        return self.nodes[num - 1]

    def compute_ordering(self):
        self.ordering = []
        stack = [1]
        visited = set()

        while len(stack) != 0:
            # num = stack.pop()
            num = stack[len(stack) - 1]
            if num not in visited:
                n = self.get_node(num)

                all_done = True
                for c in n.get_childrend_ids():
                    if c not in visited:
                        all_done = False
                        stack.append(c)

                if all_done:
                    stack.pop()
                    visited.add(num)
                    self.ordering.append(num)
            else:
                stack.pop()
                pass

    def annotate_mc(self, assumps = set()):
        for i in self.ordering:
            self.get_node(i).annotate_mc(assumps)

    def get_vars(self):
        res = set()

        for i in self.ordering:
            res.update(self.get_node(i).get_vars())

        return res

    def print_dot(self):
        print("digraph g {")
        for i in self.ordering:
            n = self.get_node(i)
            n.print_dot()
            # print(f"\"{i}\" [label=\"{i}: {n.mc}\"];")

            for c in n.get_childrend_ids():
                print(f"\"{i}\" -> \"{c}\";")

        print("}")

def from_file(path):
    res = dDNNF()

    with open(path, 'r') as file:
        for line in file:
            line = line.strip()

            if line.startswith('o '):
                num = int(line.split(' ')[1])
                res.add_node(num, OrNode)

            elif line.startswith('a '):
                num = int(line.split(' ')[1])
                res.add_node(num, AndNode)

            elif line.startswith('u '):
                num = int(line.split(' ')[1])
                res.add_node(num, UnaryNode)

            elif line.startswith('t '):
                num = int(line.split(' ')[1])
                res.add_node(num, TrueNode)

            elif line.startswith('f '):
                num = int(line.split(' ')[1])
                res.add_node(num, FalseNode)

            elif not line.startswith('c '):
                tmp = line.split(';')
                p1 = tmp[0].split(' ')

                src = int(p1[0])
                dst = int(p1[1])

                cnst = [int(x) for x in p1[2:] if x != '' and x != '0']
                free = []

                if len(tmp) == 2:
                    p2 = tmp[1].split(' ')
                    free = [int(x) for x in p2 if x != '' and x != '0']

                src = res.get_node(src)
                dst = res.get_node(dst)

                src.add_child(dst, cnst, free)

    res.compute_ordering()

    return res

File: ddnnf/src/ddnnf.jl
struct Edge
    b_lit :: Int64
    e_lit :: Int64
    b_free :: Int64
    e_free :: Int64

    child :: Int64
end

mutable struct AndNode
    children :: Vector{Edge}
    mc :: BigInt
end

mutable struct OrNode
    children :: Vector{Edge}
    mc :: BigInt
end

mutable struct UnaryNode
    child :: Union{Nothing, Edge}
    mc :: BigInt
end

struct TrueNode
end

struct FalseNode
end

const Node = Union{AndNode, OrNode, UnaryNode, TrueNode, FalseNode}

mutable struct DDNNF
    nodes :: Vector{Node}
    edges :: Vector{Edge}
    literals :: Vector{Lit}
    freev :: Vector{Var}
    ordering :: Vector{Int64}
    assumps :: Set{Lit}
end

function get_children(n :: AndNode)
    return n.children
end

function get_children(n :: OrNode)
    return n.children
end

function get_children(n :: UnaryNode)
    return [n.child]
end

function get_children(n :: TrueNode)
    return []
end

function get_children(n :: FalseNode)
    return []
end

function get_children(nnf :: DDNNF, i :: Int64)
    return get_children(nnf.nodes[i])
end

function add_edge(n :: AndNode, e :: Edge)
    push!(n.children, e)
end

function add_edge(n :: OrNode, e :: Edge)
    push!(n.children, e)
end

function add_edge(n :: UnaryNode, e :: Edge)
    n.child = e
end

function add_child(nnf :: DDNNF, src :: Int64, dst :: Int64, cnst :: AbstractArray{Lit}, free :: AbstractArray{Var})
    b_lit = length(nnf.literals) + 1
    for a in cnst
        push!(nnf.literals, a)
    end
    e_lit = length(nnf.literals)

    b_free = length(nnf.freev) + 1
    for a in free
        push!(nnf.freev, a)
    end
    e_free = length(nnf.freev)

    e = Edge(b_lit, e_lit, b_free, e_free, dst)
    add_edge(nnf.nodes[src], e)
end

function ddnnf_from_file(path :: String, prj = false, pvar = Set{Var}())
    res = DDNNF([], [], [], [], [], Set{Lit}())

    for line in eachline(path)
        if startswith(line, "o ")
            push!(res.nodes, OrNode([], 0))
        elseif startswith(line, "a ")
            push!(res.nodes, AndNode([], 0))
        elseif startswith(line, "u ")
            push!(res.nodes, UnaryNode(nothing, 0))
        elseif startswith(line, "t ")
            push!(res.nodes, TrueNode())
        elseif startswith(line, "f ")
            push!(res.nodes, FalseNode())
        elseif ! startswith(line, "c ")
            tmp = map(strip, split(line, ";"))
            p1 = split(tmp[1], " ")

            toint(x) = Base.parse(Int64, x)
            nonzero(x) = x != 0
            keep(x) = (! prj ) || (mkVar(x) in pvar)

            src = toint(p1[1])
            dst = toint(p1[2])

            cnst = filter(keep, map(mkLit, filter(nonzero, map(toint, p1[3:end]))))
            free = Vector{Var}()

            if length(tmp) == 2
                free = filter(keep, map(mkVar, filter(nonzero, map(toint, split(tmp[2], " ")))))
            end

            add_child(res, src, dst, cnst, free)
        end
    end

    compute_ordering(res)

    return res
end

function compute_ordering(nnf :: DDNNF)
    empty!(nnf.ordering)
    stack = [Int64(1)]
    visited = Set{Int64}()

    while length(stack) != 0
        num = stack[length(stack)]
        if ! (num in visited)
            getid(e :: Edge) = e.child

            all_done = true
            for c in map(getid, get_children(nnf.nodes[num]))
                if ! (c in visited)
                    all_done = false
                    push!(stack, c)
                end
            end

            if all_done
                pop!(stack)
                push!(visited, num)
                push!(nnf.ordering, num)
            end
        else
            pop!(stack)
        end
    end
end

get_free_vars(nnf :: DDNNF, i :: Edge) = @view nnf.freev[i.b_free : i.e_free]
get_literals(nnf :: DDNNF, i :: Edge) = @view nnf.literals[i.b_lit : i.e_lit]

# incomplete
function compute_free_var(nnf :: DDNNF)
    println("WARNING: usage of incomplete function compute_free_var")
    fv = Vector{Set{Var}}()
    for i in 1:length(nnf.nodes)
        push!(fv, Set{Var}())
    end

    for i in nnf.ordering
        for c in get_children(nnf, i)
            for l in nnf.literals[c.b_lit : c.e_lit]
                push!(fv[i], mkVar(l))
            end

            for v in fv[c.child]
                push!(fv[i], v)
            end
        end
    end

    m = 0
    for i in nnf.ordering
        if isa(nnf.nodes[i], UnaryNode)
            for c in get_children(nnf, i)
                println("error")
            end
        elseif isa(nnf.nodes[i], OrNode)
            for c in get_children(nnf, i)
                t2 = setdiff(fv[i], map(mkVar, nnf.literals[c.b_lit : c.e_lit]))
                setdiff!(t2, fv[c.child])
                t3 = Set(nnf.freev[c.b_free : c.e_free])

                if t2 != t3
                    m += 1
                    print("(", i, " -> ", c.child, ") ")
                    for i in t2
                        print(mkReadable(i), " ")
                    end
                    print(":")
                    for i in t3
                        print(mkReadable(i), " ")
                    end
                    print("\n   ")
                    for i in fv[i]
                        print(mkReadable(i), " ")
                    end
                    print("\n   ")
                    for i in fv[c.child]
                        print(mkReadable(i), " ")
                    end
                    print("\n   ")
                    for i in nnf.literals[c.b_lit : c.e_lit]
                        print(mkReadable(i), " ")
                    end
                    println("")
                end
            end
        end
    end
    println("m ", m)
end

